xen/arch/arm/coproc/plat/gx6xxx.c

```diff
@@ -0,0 +1,303 @@
+
+    vcoproc->priv = xzalloc(struct vgx6xxx_info);
+    if ( !vcoproc->priv )
+    {
+        dev_err(coproc->dev, "failed to allocate vcoproc private data\n");
+        return -ENOMEM;
+    }
+
+    register_mmio_handler(d, &gx6xxx_mmio_handler,
+                          mmio->addr, mmio->size, mmio);
+
+    return 0;
+}
+
+static void gx6xxx_vcoproc_deinit(struct domain *d,
+                                  struct vcoproc_instance *vcoproc)
+{
+    xfree(vcoproc->priv);
+}
+
+static const struct coproc_ops gx6xxx_vcoproc_ops = {
+    .vcoproc_init        = gx6xxx_vcoproc_init,
+    .vcoproc_deinit      = gx6xxx_vcoproc_deinit,
+    .ctx_switch_from     = gx6xxx_ctx_switch_from,
+    .ctx_switch_to       = gx6xxx_ctx_switch_to,
+};
+
+static int gx6xxx_dt_probe(struct platform_device *pdev)
+{
+    struct coproc_device *coproc;
+    struct device *dev = &pdev->dev;
+    struct gx6xxx_info *info;
+    char *reg_base;
+    int ret;
+
+    coproc = coproc_alloc(pdev, &gx6xxx_vcoproc_ops);
+    if ( IS_ERR_OR_NULL(coproc) )
+        return PTR_ERR(coproc);
+
+    if ( (coproc->num_irqs != 1) || (coproc->num_mmios != 1) )
+    {
+        dev_err(dev, "wrong number of IRQs/MMIOs\n");
+        ret = -EINVAL;
+        goto out_release_coproc;
+    }
+    coproc->priv = xzalloc(struct gx6xxx_info);
+    if ( !coproc->priv )
+    {
+        dev_err(dev, "failed to allocate coproc private data\n");
+        ret = -ENOMEM;
+        goto out_release_priv;
+    }
+    info = (struct gx6xxx_info *)coproc->priv;
+    reg_base = (char *)coproc->mmios[0].base;
+    info->reg_irq_status = (uint32_t *)(reg_base + RGXFW_CR_IRQ_STATUS);
+    info->reg_irq_clear = (uint32_t *)(reg_base + RGXFW_CR_IRQ_CLEAR);
+
+    ret = request_irq(coproc->irqs[0], IRQF_SHARED,
+                      gx6xxx_irq_handler, "GPU GX6xxx irq", coproc);
+    if ( ret )
+    {
+        dev_err(dev, "failed to request irq (%u)\n", coproc->irqs[0]);
+        goto out_release_irqs;
+    }
+
+    ret = coproc_register(coproc);
+    if ( ret )
+    {
+        dev_err(dev, "failed to register coproc (%d)\n", ret);
+        goto out_release_irqs;
+    }
+
+    return 0;
+
+out_release_irqs:
+    release_irq(coproc->irqs[0], coproc);
+out_release_priv:
+    xfree(coproc->priv);
+out_release_coproc:
+    coproc_release(coproc);
+    return ret;
+}
+
+static __init int gx6xxx_init(struct dt_device_node *dev, const void *data)
+{
+    int ret;
+
+    dt_device_set_used_by(dev, DOMID_XEN);
+
+    ret = gx6xxx_dt_probe(dev);
+    if ( ret )
+        return ret;
+
+    return 0;
+}
+
+static const struct dt_device_match gx6xxx_dt_match[] __initconst =
+{
+    DT_MATCH_GX6XXX,
+    { /* sentinel */ },
+};
+
+DT_DEVICE_START(coproc_gpu_gx6xxx, "COPROC_GPU_GX6XXX", DEVICE_COPROC)
+    .dt_match = gx6xxx_dt_match,
+    .init = gx6xxx_init,
+DT_DEVICE_END
+
+/*
+ * Local variables:
+ * mode: C
+ * c-file-style: "BSD"
+ * c-basic-offset: 4
+ * indent-tabs-mode: nil
+ * End:
+ */
```

Comment:
General comments:
1. There are places where you don't follow XEN code-style.
2. Always check for NULL before dereferencing passed to function pointers. Especially when function is global and may be called from different places in XEN.